Fix gradcheck failing for OpInfo with TensorList inputs

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53418 Fix gradcheck failing for OpInfo with TensorList inputs**

This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

heitorschueroff committed:
Fix gradcheck failing for OpInfo with TensorList inputs

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix gradcheck failing for OpInfo with TensorList inputs"


This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix gradcheck failing for OpInfo with TensorList inputs"


This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix gradcheck failing for OpInfo with TensorList inputs"


This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix gradcheck failing for OpInfo with TensorList inputs"


This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

[ghstack-poisoned]

albanD commented on test/test_ops.py:
```diff
@@ -68,7 +68,8 @@ class TestGradients(TestCase):
         @wraps(inplace_variant)
         def _fn(t, *args, **kwargs):
-            return inplace_variant(t.clone(), *args, **kwargs)
+            inputs = t.clone() if isinstance(t, torch.Tensor) else [arg.clone() for arg in t]
```
nit: is it important that inputs is a list instead of a tuple here?

heitorschueroff replied:
Nope, I made it a tuple. Thanks!

albanD reviewed (approved):
lgtm

heitorschueroff committed:
Update on "Fix gradcheck failing for OpInfo with TensorList inputs"


This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix gradcheck failing for OpInfo with TensorList inputs"


This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -85,13 +85,16 @@ def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=N
         self.output_process_fn_grad = output_process_fn_grad
 
+    def _is_tensorlist(self, arg):
+        return isinstance(arg, (tuple, list)) and len(arg) > 0 and isinstance(arg[0], torch.Tensor)
```
This is probably fine. Just fyi I recently wrote a similar function:

https://github.com/pytorch/pytorch/blob/2d36b30a8c35cc5640e1c735583c105b7037f9ca/test/test_ops.py#L466

heitorschueroff replied:
I moved this method to testing._internal.common_utils to be reused in common_methods_invocation and test_ops

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -180,7 +183,7 @@ def __init__(self,
                                                # Ex: ['aten::add', 'aten::mm'], defaults to an empty list
                                                # Note: currently no ops use fusible nodes
-                 output_func=lambda x: x,  # fn mapping output to part that should be gradcheck'ed
+                 grad_func=lambda op, *args, **kwargs: op(*args, **kwargs),  # wrapper function for gradcheck
```
What about making this None and if None it just points to the op?

heitorschueroff replied:
It has to take op has its first argument so it can wrap the op variants as well. I gave it a default lambda as a way to document its use.

mruberry replied:
OK.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -976,81 +973,25 @@ def __init__(self,
         self.ref = ref
 
 
-class HermitianOpInfo(OpInfo):
```
I'm not sure if this PR can remove this class. cc @IvanYashchuk, you're updating this class in another PR of yours, right?

IvanYashchuk replied:
It's great that `grad_func=` replaces `HermitianOpInfo` and `TriangularOpInfo`! But it will hit a merge conflict with https://github.com/pytorch/pytorch/pull/51107
Let's wait until that PR is merged and then rebase this one.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2539,51 +2505,41 @@ def _make_tensor_helper(shape, low=None, high=None):
            skips=(
-               SkipInfo('TestCommon', 'test_variant_consistency_jit',
-                        dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16)),
+               # This test is not working properly for TensorList inputs
```
Let's file and link an issue for this

heitorschueroff replied:
Created issue here https://github.com/pytorch/pytorch/issues/53906

mruberry commented:
I really like this change and the design looks good. I've asked @IvanYashchuk to weigh in on removing the two OpInfo classes for linalg operations. Although this PR can replace them in the current build, I think he may have plans for them.

Looks like the test failures are just mypy issues.

IvanYashchuk commented:
This PR resolves https://github.com/pytorch/pytorch/issues/50837.

heitorschueroff committed:
Update on "Fix gradcheck failing for OpInfo with TensorList inputs"


This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fix gradcheck failing for OpInfo with TensorList inputs"


This PR adds a workaround to OpInfo tests for ops that take TensoList inputs which were failing gradcheck (see https://github.com/pytorch/pytorch/issues/51996). It also updates the *stack ops to pass in a list of Tensors as input to SampleInput to demonstrate that the tests now pass without the need for defining a lambda in op.

Differential Revision: [D26860788](https://our.internmc.facebook.com/intern/diff/D26860788)

[ghstack-poisoned]

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -575,26 +606,24 @@ def sample_inputs_div(self, device, dtype, requires_grad, rounding_mode=None):
-                        make_tensor((S, S), device, dtype,
-                                    low=None, high=None,
-                                    requires_grad=requires_grad)), kwargs=dict(idx=0)),)
+    tensors = [make_tensor((S, S), device, dtype,
```
For now, TensorList inputs must be List[Tensor]. @heitorschueroff find out why Tuple[Tensor] is not supported here.
